### PR TITLE
Lberg/pipenv

### PR DESCRIPTION
**Pull request by lucabergamini**
Add pipenv as dependencies manager

### CODE REVIEW

File: .github/workflows/docs.yml
Comment (giacomodabisias) [inline .github/workflows/docs.yml:26]: what is this last line for?

File: .github/workflows/docs.yml
Comment (giacomodabisias) [inline .github/workflows/docs.yml:25]: Always use `pipenv sync` to make sure that you install packages from the lock file.
No need for `--dev` since you are not developing here

File: l5kit/.mypy.ini
Comment (giacomodabisias) [inline l5kit/.mypy.ini:9]: missing new line

File: l5kit/Pipfile
Comment (giacomodabisias) [inline l5kit/Pipfile:42]: no need for this since you pinned the black version and nothing else looks like a pre-release

File: l5kit/Pipfile
Comment (giacomodabisias) [inline l5kit/Pipfile:36]: This seems dangerous. Was there any incompatible change lately? If yes please add a comment

**Comment by giacomodabisias** [comment]
Small NIT: Always add a description to a PR to give a high overview of what is going on :)

File: .github/workflows/docs.yml
Comment (lucabergamini) [reply .github/workflows/docs.yml]: we read the version from the `__init__` in l5kit, so we need it to be installed here..

File: .github/workflows/docs.yml
Comment (lucabergamini) [reply .github/workflows/docs.yml]: Currently our doc tools are installed as `dev` dependencies so they need `--dev`, but maybe that's not a best practice?

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile:39]: It's autogenerated, should I still remove it?

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile:36]: No, but macOS has already the 7.17 while linux doesn't. If I generated the lock from MacOS without capping the version CI workers won't be able to install from the lock..

File: .github/workflows/docs.yml
Comment (giacomodabisias) [reply .github/workflows/docs.yml]: not sure I understand. What is this installing?

File: l5kit/Pipfile
Comment (giacomodabisias) [reply l5kit/Pipfile:36]: You mean that the latest versions here https://pypi.org/project/ipython/#history are only for Mac? That seems very strange. I would assume that thing are always first available for linux and later for mac, if  this actually happens.

File: l5kit/Pipfile
Comment (giacomodabisias) [reply l5kit/Pipfile:39]: How are you autogenerating the Pipfile?

File: .github/workflows/docs.yml
Comment (lucabergamini) [reply .github/workflows/docs.yml]: l5kit itself from the `setup.py`

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile:39]: No sorry I mean those lines were autogenerated after installing black.

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile:36]: I've checked now and it seems like that issue has been fixed. Let me try to remove the line and see if it passes on CI machines

File: .github/workflows/docs.yml
Comment (giacomodabisias) [reply .github/workflows/docs.yml]: but if you add l5kit to the pipfile there is no need for this, right?

File: .github/workflows/docs.yml
Comment (giacomodabisias) [reply .github/workflows/docs.yml]: docs are not part of dev :)
Dev packages are packages needed for development and documentation is not part of that

File: l5kit/Pipfile
Comment (giacomodabisias) [reply l5kit/Pipfile:39]: no need for that since you specify manually the black version as a pre release

File: l5kit/Pipfile
Comment (giacomodabisias) [reply l5kit/Pipfile:36]: nice :)

File: .github/workflows/docs.yml
Comment (lucabergamini) [reply .github/workflows/docs.yml]: I agree, but they are also not part of the library itself right? Should they be install together with it regardless?

File: .github/workflows/docs.yml
Comment (lucabergamini) [reply .github/workflows/docs.yml]: ok that makes sense, should I add it in editable mode then?

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile]: checked again, newer ipython are [only 3.7+](https://ipython.readthedocs.io/en/stable/whatsnew/version7.html) . I can generate the lock from a 3.6 python to ensure the locked version is 3.6-compatible, is that fine with you? The risk there is obviously someone updating the lock in the future from 3.8  or 3.7.

Alternatively, we could drop support for 3.6

File: .github/workflows/docs.yml
Comment (giacomodabisias) [reply .github/workflows/docs.yml]: yep, I would add it there so that pipenv manages everything. There should't be the need to run any pip command

File: .github/workflows/docs.yml
Comment (giacomodabisias) [reply .github/workflows/docs.yml]: the dev packages should be installed when you want to do development :) In CI there is no development so there should be no --dev. Whatever is needed to produce some artifacts or to test something should not be dev

File: l5kit/Pipfile
Comment (giacomodabisias) [reply l5kit/Pipfile:36]: why is ipython in the pipfile? 
is there a dependency on it?

Comment (lucabergamini) [reply l5kit/Pipfile:36]: jupyter has a dependency on ipython. If I let pipenv to manage it it will have a different behaviour depending on the python version you use to generate the lock:
- python 3.6 -> lock 7.16
- python 3.7, 3.8 -> lock 7.18 or the most recent one

The issue here is that ipython has dropped support for 3.6 between two versions, what is the best practise here?

**Comment by Jimexist** [comment]
I wonder if we can use [Poetry](https://python-poetry.org) instead? Since there's already pyproject.toml file in place.

**Comment by lucabergamini** [comment]
> I wonder if we can use [Poetry](https://python-poetry.org) instead? Since there's already pyproject.toml file in place.

Big fan of Poetry here (there is even a branch which supports it already), but for internal reason we have to use pipenv...

**Comment by giacomodabisias** [comment]
@Jimexist  Poetry could be the future of dependency management, but currently it is still at version 1.0.10. 
i'd wait a bit longer before using something which is so fresh and use something more reliable and stable in the meantime (not saying that pipenv is perfect, but it has been around for longer)

File: l5kit/Pipfile
Comment (giacomodabisias) [reply l5kit/Pipfile:36]: jupyter should resolve its dependencies on its own. II don't think you need to add the internal dependencies.
In general I don't think its a great idea to support several python versions. What is the advantage? You can just stick to one version.

**Comment by lucabergamini** [comment]
Also, Poetry doesn't support `setup.py` which is still the only supported way of installing a package from pip (this will likely change in the future)

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile:36]: > jupyter should resolve its dependencies on its own. II don't think you need to add the internal dependencies.

It's a locking issue here. If we lock from 3.8, ipython will be locked at 7.18 which is not available on 3.6 in CI.



> In general I don't think it's a great idea to support several python versions. What is the advantage? You can just stick to one version.

Well I guess one advantage is that if you're on ubuntu 18.04LTS (default python is still 3.6.8 there) you don't have to install a new interpreter just to use our package. Apart from that, I don't see any particular issues in dropping support for 3.6 and 3.7. It will also make maintenance easier. So should we just aim to support 3.8 in the end @giacomodabisias ?

**Comment by Jimexist** [comment]
> Also, Poetry doesn't support `setup.py` which is still the only supported way of installing a package from pip (this will likely change in the future)

agree that in this case Pipenv is a better choice.

File: l5kit/setup.py
Comment (giacomodabisias) [inline l5kit/setup.py:34]: why do you need `mypy>=0.782` and "pytest-cov>=2.10.0" and "flake8>=3.8.3" and "isort>=5.0.4" ?
These are dev packages and you should be able to always use the latest version right? 
Is there any specific thing happening with the pinned versions?

Comment (lucabergamini) [reply l5kit/setup.py:34]: There are issues happening with previous versions (so for example mypy 0.77). I guess this is not really necessary though, as the behaviour should be the same as `*`. I'll remove it

**Comment by Jimexist** [comment]
i am all for dropping py3.6

**Comment by lucabergamini** [comment]
> i am all for dropping py3.6

Already included in this PR :) We'll stop testing it in CI :)

File: .github/workflows/l5kit-lock.yml
Comment (giacomodabisias) [inline .github/workflows/l5kit-lock.yml:26]: is there no need for a `wait` here since all steps run in parallel?

File: .github/workflows/l5kit-lock.yml
Comment (giacomodabisias) [inline .github/workflows/l5kit-lock.yml:16]: can we indent all steps for readability?

File: README.md
Comment (giacomodabisias) [inline README.md:125]:
```suggestion
pipenv sync --dev
```
for readability

File: .github/workflows/l5kit-lock.yml
Comment (giacomodabisias) [inline .github/workflows/l5kit-lock.yml:26]:
```suggestion
        pipenv sync --dev
```
for reaedability

File: README.md
Comment (giacomodabisias) [inline README.md:127]:
```suggestion
This will install all dependencies (`--ddev` includes dev-packages too) from the lock file.
```

File: l5kit/Pipfile
Comment (giacomodabisias) [inline l5kit/Pipfile:36]: I'd move this to thee top of the list

Comment (giacomodabisias) [inline l5kit/Pipfile:36]:
```suggestion
l5kit = {editable = true, path = "./l5kit"}
```

**Review by giacomodabisias — COMMENTED** [review]
Nice progress!
Almost there, some more minor comments

File: .github/workflows/l5kit-lock.yml
Comment (lucabergamini) [reply .github/workflows/l5kit-lock.yml:26]: As far as I know `jobs` are run in parallel in a workflow, but `steps` and `run` are sequential. I would be very surprised if that was not the case ad we've never experienced any sync issues.
https://resources.github.com/whitepapers/GitHub-Actions-Cheat-sheet/ also seems to confirm it (I haven't found a more official doc that explains it though)

Comment (giacomodabisias) [reply .github/workflows/l5kit-lock.yml]: Ah wait, I aws looking at buildkite xD never mind!

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile]: I'm not able to `pipenv lock` anymore if I add this:
`pipenv.vendor.requirementslib.exceptions.RequirementError: Failed parsing requirement from './l5kit'` 
 
Why should it point to `./l5kit` and not `.` ?

**Review by giacomodabisias — DISMISSED** [review]
Great Job :)

File: l5kit/Pipfile
Comment (lucabergamini) [reply l5kit/Pipfile]: I think [it should point](https://stackoverflow.com/questions/53378416/what-does-pipenv-install-e-do-and-how-to-use-it) to where the `setup.py` is (otherwise pip doesn't know how to build it). If that's the case it makes sense to leave it as `.` here, what do you think @giacomodabisias

Comment (giacomodabisias) [reply l5kit/Pipfile]: Ah your pipfile is in the L5kit folder, I see. Usually you have the pipfile and pipfile.lock one folder up, since you might create new packages in the same repo and you can use the pipfile to manage all of them if they work together.
I suggest moving the files up one folder to the main folder of the repo

Comment (lucabergamini) [reply l5kit/Pipfile]: mmm...given that this is not the case for this repo (not planned anyway) I would leave it there. The reason is also that all other configuration files (and `setup.py` itself) are already placed there. Moving only this one one level up would create some confusion about where to run what imo (e.g. install and add packages vs running tests, lint or installing from `setup.py`).

Comment (giacomodabisias) [reply l5kit/Pipfile]: up to you :) I am fine also with the current structure.
Notice though that there would be no confusion since tests, linting etc..are all independent from the file location.
Setup.py should be in the package as it is now and should not be moved :)

Comment (lucabergamini) [reply l5kit/Pipfile]: > Notice though that there would be no confusion since tests, linting etc..are all independent from the file location.

They still need to read their config files though, otherwise they run with default configs.

I'll leave it as it is for now but I'll consider it for the future :) Can I have your final blessing @giacomodabisias 😄 ?

Comment (giacomodabisias) [reply l5kit/Pipfile]: sure, go for it

**Review by yye-lyft — APPROVED** [review]

**Comment by Jimexist** [comment]
thanks for the change. do we know if we want to cut a new release after this?

**Comment by lucabergamini** [comment]
> thanks for the change. do we know if we want to cut a new release after this?

Likely after #150 and #151 are merge, hopefully by middle of the next week we should have a new release tested :)

**Pull request merged**
